SilverStripe 4 upgrade

3Dgoo opened the pull request:
This pull request is for upgrading this module to work for SilverStripe 4. This is a work in progress and is not ready to be merged in yet.

Work still to be done:

- [x] Test tests
- [x] Work out how to correctly extend module in SS4
- [x] Update documentation
- [x] Test search auto redirect functionality
- [x] Check requirements are correct
- [x] Update Travis config to test with SS4 and the supported PHP versions
- [x] Change src/* folder names to start with a capital letter and update documentation to reflect that change

3Dgoo committed:
Updating code for SilverStripe 4. Still work to be done. Update panel needs fixing to display when needed. Search needs fixing. Search pagination does not work. Initial search does not work. Documentation needs updating. Requirements need updating.

3Dgoo committed:
Fixing ss4 dashboard search. Fixing dashboard being default panel.

3Dgoo committed:
Adding use to tests. Fixing update panel. Fixing update panel styling.

3Dgoo committed:
Update compose requirement to framework 4.0

3Dgoo committed:
Tweaking ss4 requirement

Turnerj commented on src/panels/UpdatePanel.php:
```diff
@@ -59,23 +71,22 @@ public function getContactContent()
         }
 
-        $versions = explode(', ', Injector::inst()->get('LeftAndMain')->CMSVersion());
+        $versions = explode(', ', Injector::inst()->get('SilverStripe\Admin\LeftAndMain')->CMSVersion());
```
We need to make sure we address the issues raised in #107 here too. I see you changed it to use "CMS: " but was that all we really needed to change to fix this?

3Dgoo replied:
The reason I changed it to use CMS instead of Framework is because retrieving Framework in 4.0 came up with null for me. This change was purely to get things working. Do you think we should leave it as changed to check the CMS version? Or investigate why fetching Framework didn't work?

Turnerj replied:
I think from @aNickzz 's description, it sounds like the SilverStripe CMS module actually intentionally sets the Framework version to nothing. It might be exactly the same thing you saw in SS4.

I'm happy to just use the CMS version but I guess one thing we should probably do is a few more safety checks around the version number in case from some reason "CMS" is blank too.

3Dgoo commented on _config/config.yml:
```diff
@@ -1,10 +1,17 @@
+  Psr\SimpleCache\CacheInterface.myCache:
+    factory: SilverStripe\Core\Cache\CacheFactory
+    constructor:
+      namespace: "myCache"
```
@Turnerj I don't know if this is the correct way to call the cache. This is from the documentation. I guess we should change _myCache_ to something more meaningful like _plastykDashboardCache_. What do you think?

Turnerj replied:
Yeah, I agree. I saw your other comments where there are currently references to _myCache_ and I'm thinking like you are that we should change it to something more meaningful.

I think _plastykDashboardCache_ would work. I was wondering whether having "cache" in the namespace name was a little redundant but from a quick look through the documentation, it seems to be what they preference.

3Dgoo commented on src/model/UpdateVersionList.php:
```diff
@@ -52,8 +58,8 @@ public function sortByVersion($direction = 'ASC')
     {
-        $versionListCache = SS_Cache::factory('VersionList');
-        $result = $versionListCache->load('PackagistVersions');
+        $versionListCache = Injector::inst()->get(CacheInterface::class . '.myCache');
```
@Turnerj  _myCache_ is also referenced here.

3Dgoo commented on src/panels/UpdatePanel.php:
```diff
@@ -59,23 +71,22 @@ public function getContactContent()
-        $result = $updatePanelCache->load('CurrentSilverStripeVersion');
-        if ($result) {
-            return UpdateVersion::from_version_string($result);
+        $updatePanelCache = Injector::inst()->get(CacheInterface::class . '.myCache');
```
@Turnerj _myCache_ is also referenced here.

3Dgoo committed:
Fixing search issues. Javascript encoding problem with multiple words. Fixing no exact matches error.

3Dgoo committed:
Changing caching from myCache to plastykDashboardCache.

3Dgoo committed:
Adding namespacing to tests.

3Dgoo committed:
Fixing scrutinizer path.

3Dgoo committed:
Fixing travis testing to target correct versions of php and framework.

3Dgoo committed:
Travis changes.

Turnerj commented on .travis.yml:
```diff
@@ -1,28 +1,34 @@
+  - composer validate
+  - composer require --no-update silverstripe/installer:4.0.x-dev
+  - if [[ $DB == PGSQL ]]; then composer require --no-update silverstripe/postgresql:2.0.x-dev; fi
+  - composer install --prefer-dist --no-interaction --no-progress --no-suggest --optimize-autoloader --verbose --profile
```
May need some more tweaks, Travis is saying our requirements could not be resolved to an installable set of packages.
https://travis-ci.org/plastyk/silverstripe-dashboard/jobs/347232337

3Dgoo committed:
Fixing tests to call namespaced Member and Group objects.

3Dgoo committed:
Trying to fix travis.

3Dgoo committed:
Travis.

3Dgoo committed:
Travis.

3Dgoo committed:
Composer tweaks.

3Dgoo committed:
Composer tweak.

3Dgoo committed:
Revert "Composer tweak."

This reverts commit 12f5bfc0749d3a8ba3bb7ddbb274ec535d7e75b2.

Turnerj committed:
Update minimum stability

Matches what some of the other SS modules have configured, should resolve Travis issues

Turnerj committed:
Fixed typo

Doh. This is why I shouldn't edit things directly through the Github UI

Turnerj commented:
Good news, I resolved our initial issue with Travis. I saw that the different SS modules all specify a minimum stability of "dev" in Composer, copying that solved that problem. Also seems to have addressed Scrutinizer's issue (though looks like we have a lot to fix/check).

We have a couple of different problems now with Travis:
- It can't connect to the MySQL server for running the tests (likely another minor configuration thing)
- Based on our Travis config, it is performing some linting as part of a test. The linting is failing.

We have to fix the first one and we can optionally fix the second if we want linting performed by Travis (could be useful).

I did notice with the Travis config that you've setup, it has a few lines regarding code coverage. I'm keen to finish the configuration for that so we can display our code coverage on the readme.

Turnerj commented:
Progress is slow on fixing Travis at the moment. This is what Travis says: 
```(HY000/1045): Access denied for user ''@'localhost' (using password: NO)```

Looking it up, Travis for MySQL has two users set up, "root" and "travis". Strange thing is, I can't find any reference to these being defined in any of the other SS modules so I don't think we should be defining it ourselves.

Turnerj committed:
Constant logging for Travis

Turnerj commented:
OK, maybe a small amount of progress. From [a commit](https://github.com/silverstripe/silverstripe-framework/commit/6af1e5bf0daafea3a833f3795b95243d50adc2dd) I found in Framework's `.travis.yml` file, you can set environment variables in Travis (and [SilverStripe may read them](https://docs.silverstripe.org/en/4/getting_started/environment_management/)). After many failed attempts, I've managed to get [a different error](
https://travis-ci.org/plastyk/silverstripe-dashboard/jobs/348908261#L661).

While this may seem promising, I don't think its the right way to do it. I think this is the case as to get it to work, I needed to make [environment variables display in Travis](https://travis-ci.org/plastyk/silverstripe-dashboard/jobs/348908261#L434). Whereas, if we look at Framework, [it doesn't have that](https://travis-ci.org/silverstripe/silverstripe-framework/jobs/306225543#L1246) but its tests work fine.

I think we should probably query someone of the SilverStripe slack channel to see if they can shed any light on this,

3Dgoo committed:
Fixing scrutinizer issue: The variable $result does not seem to be defined for all execution paths leading up to this point.

3Dgoo committed:
Scrutinizer minor fixes.

3Dgoo committed:
Making DashboardPanel and DashboardSearchResultPanel abstract classes.

3Dgoo reviewed (commented):
@Turnerj - does this last pull request look OK to you?

3Dgoo committed:
Documentation updates. Updating screenshot images. Updating parts of documentation. More documentation still needs updating. Update file search on new different file fields.

3Dgoo committed:
Documentation updates.

3Dgoo committed:
Exposing assets

Adding expose setting to composer.json to make css, images and javascript accessible.

3Dgoo committed:
Moving module to vendor directory

3Dgoo committed:
Updating QuickLinksPanelExtension example

3Dgoo committed:
Fixing search panel documentation example

Changing Location to Property in search panel example documentation

Bob-Plastyk committed:
Fixing documentation to not reference location. Fixing documentation to capitalise src sub folders.

Bob-Plastyk committed:
Changing src sub folder capitalisation

3Dgoo committed:
Changing setting class names of search to us ::class

3Dgoo committed:
Fixing link routing JavaScript.

3Dgoo committed:
Minor test tweaks.

3Dgoo committed:
Merge branch 'master' into ss4

# Conflicts:
#	src/Panels/UpdatePanel.php

3Dgoo committed:
Changing FunctionalTest to SapphireTest.

3Dgoo committed:
Setting different SilverStripe versions for the different tests.

3Dgoo committed:
no message

3Dgoo committed:
Calling DashboardAdmin::create in test setUp to instantiate a controller.

3Dgoo committed:
Trying to fix travis tests with useDatabase setting.

3Dgoo committed:
Changing travis settings to try to fix travis tests.

3Dgoo committed:
Travis tweak.

3Dgoo committed:
Little travis fix.

3Dgoo committed:
Adding tests directory to travis file.

3Dgoo committed:
Travis tweak.

3Dgoo committed:
Let's see if we need this anymore.

3Dgoo committed:
Removing things that don't need to be here.

3Dgoo committed:
Adding code sniffer rules. Apply code sniffer fixes. Changing array syntax.

3Dgoo committed:
Fixing a part of the search.

3Dgoo committed:
Fixing search get plural / singular names.

3Dgoo committed:
Fix to internal links loading pages / files incorrectly by forcing the link to reload the whole page.

3Dgoo commented:
@Turnerj do you want to check that this is ready for release?

Turnerj reviewed (commented):
There are a few think at a quick glance over that might need adjustment or explanation. Github is being weird and not letting me do inline reviews on the files (saying the commit is not part of the PR)...

So three of the documentation pages, there are changes which I think are a mistake. `Permission::check()` runs on who is logged in, not on the `$member` parameter like `Permission::checkMember()` does.
See: "customisation-extra-search-panels.md", "customisation-new-panels.md" and "customisation-quicklinks.md".

Besides that, the only thing that stands out for me is `DashboardPanel` and `UpdateVersion` - both are now `DataObject` and `DashboardPanel` is no longer abstract. I do recall us discussing about this before but I don't recall what it was about those two things that was the problem.

Neither object writes to the DB so having it as a `DataObject` seems weird. Removing abstract from `DashboardPanel` makes me think I can just construct that object directly which we really shouldn't be able to.

Can you recall what our reasons were for that?

Turnerj commented:
What I am thinking though is that we merge our changes into master soon as we are like 95% of the way there. We then can break down anything else for SS4 into smaller chunks.

Just want to avoid introducing a breaking change when releasing SS4 so we don't have to skip from V2 to V3 too fast.

Turnerj commented:
One little thing we should probably do is actually put the SS version support in our readme so people know what version it supports by looking at the project homepage.

3Dgoo committed:
Adding SilverStripe 4.0 requirement in read me

3Dgoo commented:
I have added the SS version in the readme. 

As for the other issues you raised, I have a feeling you're viewing an old version.

`DashboardPanel` and `UpdateVersion` and both abstract and do not extend `DataObject`.
All the documentation calls `Permission::checkMember()`. 

Have another look and let me know if there are any other issues you see.

Turnerj commented:
@3Dgoo Thanks for all your hard work on this, I really appreciate the effort you've put in.
